torch/_inductor/codegen/cuda/cuda_scheduling.py

```diff
@@ -1,15 +1,199 @@
+
+    Args:
+        cuda_template_buffer : A CUDATemplateBuffer object representing the CUDA template and it's result buffer
+        epilogue_nodes : List[ir.Buffer]: The list of already fused epilogue nodes.
+        additional_node: The ir.Buffer node to be checked if it can be fused with the epilogue.
+    Returns:
+    - bool: True if the given node can be fused with the epilogue, False otherwise.
+
+    """
+    if not cuda_template_buffer.template.can_fuse_epilogue:
+        # The used GEMM op does not support fusing epilogues
+        return False
+
+    if not isinstance(additional_node, ComputedBuffer):
+        return False
+    if not isinstance(additional_node.data, Pointwise):
+        return False
+
+    # We can fuse a Pointwise op that depends on the last fused epilogue node
+    # if any. If there is no epilogue node yet, it needs to depend on the template
+    # node
+    node_name = additional_node.name if additional_node.name is not None else additional_node.data.name  # type: ignore[attr-defined]
+    if node_name is None:
+        return False
+
+    if len(epilogue_nodes) == 0:
+        if cuda_template_buffer.name not in additional_node.get_read_names():
+            return False
+    else:
+        last_epilogue_node = epilogue_nodes[-1]
+        last_epilogue_name = (
+            last_epilogue_node.name
+            if last_epilogue_node.name is not None
+            else last_epilogue_node.data.name  # type: ignore[attr-defined]
+        )
+        if last_epilogue_name not in additional_node.get_read_names():
+            return False
+    if additional_node.layout != cuda_template_buffer.layout:
+        return False
+    try:
+        from torch._inductor.codegen.cuda.cutlass_epilogue_gen import (
+            CutlassEVTEpilogueArgumentFormatter,
+            CutlassEVTEpilogueTypeFormatter,
+        )
+
+        CutlassEVTEpilogueTypeFormatter.ir_to_evt_string(
+            cast(str, cuda_template_buffer.name), "anything", [additional_node]
+        )
+        CutlassEVTEpilogueArgumentFormatter.ir_to_evt_argument_string(
+            cast(str, cuda_template_buffer.name), [additional_node]
+        )
+    except NotImplementedError as e:
+        not_implemented_op = str(e)
+        if not_implemented_op.startswith("_op_"):
+            not_implemented_op = not_implemented_op[4:]
+            log.warning(
+                f"Cannot fuse epilogue node {additional_node} into {cuda_template_buffer.name}, likely due to unsupported operation: {not_implemented_op}"  # noqa: G004, B950
+            )
+            return False
+        else:
+            # Likely due to unsupported dtype.
+            log.warning(
+                f"Cannot fuse epilogue node {additional_node} into {cuda_template_buffer.name}. Reason: {not_implemented_op}"  # noqa: G004, B950
+            )
+            return False
+    return True
+
+
+class CUDASchedulerNode(SchedulerNode):
```

Comment:
`CUDASchedulerNode` is not at the same abstraction level compared to other `SchedulerNode`s defined in `scheduler.py`. It seems to me that the APIs here should be moved to `CUDAScheduling` instead.

Why not just implementing `CUDAScheduling.can_fuse()`?

Reply:
I didn't want to couple scheduler.py to cuda_scheduling ( e.g. require it to import that, which might lead to a circular import). 

These methods are actually called as instance methods on their abstract baseclass if the methods exist. ( That's happening in scheduler.py ).

And yes, the point of these classes is to specialize them for wrapping CUDATemplates. What would be the point of moving that to CUDAScheduling and hardcoding the dispatch / special case into scheduler.py?